### ISSUE TITLE
Extension filtering in No Profile Mode - questions.

### ISSUE THREAD

**Issue by nishihatapalmer** [digital-preservation/droid#570]
Trying to understand the purpose of extension filtering in no profile mode on the command line.  I can see two possibilities:

1. It's to limit the output to only the kinds of things you're interested in.  
2. It's a performance optimisation to prevent scanning things you're not interested in.

I'm currently re-implementing no profile mode to use the standard profile machinery to output to CSV on the console, or to a file, to avoid the current code duplication and re-implementation that exists between the two.  So in order to implement extension filtering, I have to be able to build it in to the profile machinery. 

My preferred choice is to use standard filter definitions to filter the output.  This will be much more flexible, as you could filter on anything, not just the file extension.  I would keep the current command line extension filtering option too, and just alias it to create an extension filter.  However - if the goal was to avoid scanning for performance reasons, this wouldn't achieve the same goal, as the filter would be applied after scanning each file, not before.

If it is for performance reasons, then it would obviously be worthwhile to explore performance differences beween no profile mode and the new version that uses profile machinery, as they will have different performance characteristics in the first place.

The final question I have, is that extension filtering for no profile mode currently only filters on top level files.  Anything inside a zip file for example won't be extension filtered.  My intention is to make extension (or other types) filtering apply to all files regardless of where they are found.  I assume that although this would be a change from the current system, it would be acceptable...?

**Comment by nishihatapalmer** [comment]
One more question on extension filtering, in relation to how it handles archive file types.

Let's say I filter a zip extension, but I've got the system set to expand zip files.   I assume that since I'm filtering zip extensions, and I find a zip file with a zip extension, the archive shouldn't be expanded.  But if I find a zip file without a zip extension, it would be expanded and processed?  The current code in no profile mode would behave this way.

In the GUI, filtering is different, in that it never prevents the scanning of files or processing of archives, it just filters the output.  Everything gets processed and recorded, but the GUI or the export can filter out what you're not interested in.  This is because you can efficiently query a database later, and the goal of filtering here is not performance enhancement, but result exploration.

I guess this means that if we are filtering on the command line for CSV output, we should stick with the current behaviour - if filtered, don't process as an archive, as there is no output for the high level archive file (so would be no parent object in the output).

**Comment by nishihatapalmer** [comment]
An update, as I've now got results filtering working on direct CSV output in the command line (not an export from a database profile).  We still need to set parent ids on filtered objects, otherwise errors happen.  For example, if I'm filtering so only files are output, folders still need ids even though they're not written out.  So the above is correct - filtering a zip extension shouldn't prevent processing the zip file contents - it should just prevent writing out the actual zip file as an identification result.  If we wanted to prevent zip file expansion, we already have other options for that.  Filtering is just about what you see in the results, not what gets processed.

**Comment by nishihatapalmer** [comment]
The one remaining way that filtering could affect what gets processed, and thus performance potentially, is if file metadata only was filtered BEFORE submitting the file for processing.  Obviously, if filtering on identifications, then the file needs to be processed before it can be filtered.  This will be a little bit more complicated, so would still like to understand the original rationale for extension filtering.

**Comment by nishihatapalmer** [comment]
OK, well... I'm implementing an IdentificationRequestFilter as well, which can filter out request metadata that exists before identification takes place.  This only include the filename, file size, last modified date and file extensions.  Turns out I can reuse most of the work I did on post identification results filtering.

So we'll get the best of both worlds: if a filter would filter the results out of a CSV output, and we can do that before identification takes place, we'll do it there, avoiding running the identification for that request.  If there are any criteria which would affect the output post-identification, then we do that as well and run identification on it.
